### PR TITLE
Make the gen-i18n script validate _t calls

### PR DESCRIPTION
**Pull request by dbkr**
And throw a massive tantrum if you've messed up your format strings.

Because broken format strings making their way into the app cause it
to throw exceptions.

### CODE REVIEW

File: scripts/gen-i18n.js [hunk 1]
```diff
@@ -73,6 +74,41 @@ function getTKey(arg) {
...
+        if (placeholder === '%') continue; // Literal % is %%
+
+        const placeholderMatch = placeholder.match(/^\((.*?)\)(.)/);
+        if (placeholderMatch === null) {
+            throw new Error("Invalid format specifier: '"+match[0]+"'");
+        }
+        if (placeholderMatch.length < 3) {
+            throw new Error("Malformed format specifier");
+        }
+        const placeHolderName = placeholderMatch[1];
```
Comment (lukebarnard1) [inline scripts/gen-i18n.js]: `placeholderX` is fine I think (lowercase "placholder")

File: scripts/gen-i18n.js [hunk 2]
```diff
@@ -73,6 +74,41 @@ function getTKey(arg) {
...
+        const placeHolderFormat = placeholderMatch[2];
+
+        if (placeHolderFormat !== 's') {
+            throw new Error(`'${placeHolderFormat}' used as format character: you probably didn't mean this`);
```
Comment (lukebarnard1) [inline scripts/gen-i18n.js]: Maybe mention that only "s" is supported? as much as I would love a "computer says 'no'"-style response :wink:

File: scripts/gen-i18n.js [hunk 3]
```diff
@@ -89,6 +125,28 @@ function getTranslationsJs(file) {
...
+                            if (node.arguments.length < 2) {
+                                throw new Error(`Placeholder found ('${placeholder}') but no substitutions given`);
+                            }
+                            const value = getObjectValue(node.arguments[1], placeholder);
+                            if (value === null) {
+                                throw new Error(`No value found for placeholder '${placeholder}'`);
+                            }
+                        }
+                    } catch (e) {
+                        console.log();
```
Comment (lukebarnard1) [inline scripts/gen-i18n.js]: Empty log?

File: scripts/gen-i18n.js [hunk 4]
```diff
@@ -89,6 +125,28 @@ function getTranslationsJs(file) {
...
+                        console.error(`ERROR: ${file}:${node.loc.start.line} ${tKey}`);
+                        console.error(e);
```
Comment (lukebarnard1) [inline scripts/gen-i18n.js]: I don't think the stack trace is useful TBH (having tested, it's only necessary to log the location of the translation string)

File: scripts/gen-i18n.js [hunk 5]
```diff
@@ -182,7 +240,9 @@ for (const tr of translatables) {
 }
 
 fs.writeFileSync(
-    "src/i18n/strings/en_EN.json",
+    OUTPUT_FILE,
     JSON.stringify(trObj, translatables.values(), 4) + "\n"
 );
 
+console.log();
```
Comment (lukebarnard1) [inline scripts/gen-i18n.js]: Empty log?

File: scripts/gen-i18n.js (hunk 3 above)
Comment (dbkr) [reply scripts/gen-i18n.js]: Ah, these are actually intentional to log blank lines which makes the output scan a bit better (I went with a separate log call vs putting a literal newline in the log string so the program looks a bit more like the output).

File: scripts/gen-i18n.js (hunk 5 above)
Comment (dbkr) [reply scripts/gen-i18n.js]: also for the newline

File: scripts/gen-i18n.js (hunk 4 above)
Comment (dbkr) [reply scripts/gen-i18n.js]: fair

File: scripts/gen-i18n.js (hunk 2 above)
Comment (dbkr) [reply scripts/gen-i18n.js]: heh, yes, fair

File: scripts/gen-i18n.js (hunk 1 above)
Comment (dbkr) [reply scripts/gen-i18n.js]: oops

File: scripts/gen-i18n.js [hunk 6]
```diff
@@ -73,6 +74,41 @@ function getTKey(arg) {
...
+    // We could make a regex that matched the full placeholder, but this
+    // would just not match invalid placeholders and so wouldn't help us
+    // detect the invalid ones.
+    // Also note that for simplicity, this just matches a % character and then
+    // anything up to the next % character (or a single %, or end of string).
+    const formatStringRe = /%([^%]+|%|$)/g;
+    const formatStrings = new Set();
+
+    let match;
+    while ( (match = formatStringRe.exec(str)) !== null) {
```
[the remark this answers is not in the document]
Comment (dbkr) [reply scripts/gen-i18n.js]: oh yeah, oops

File: scripts/gen-i18n.js [hunk 7]
```diff
@@ -85,7 +85,7 @@ function getFormatStrings(str) {
     const formatStrings = new Set();
 
     let match;
-    while ( (match = formatStringRe.exec(str)) !== null) {
+    while ( (match = formatStringRe.exec(str) ) !== null) {
```
Comment (lukebarnard1) [inline scripts/gen-i18n.js]: Still uneven :p

File: scripts/gen-i18n.js (hunk 3 above)
Comment (lukebarnard1) [reply scripts/gen-i18n.js]: Fair enough!

**Comment by lukebarnard1** [comment]
https://travis-ci.org/matrix-org/matrix-react-sdk/builds/291531432?utm_source=github_status&utm_medium=notification is a very strange failure mode. Seems totally unrelated especially as the tests don't run this script.

**Pull request merged**